core/include/vecmem/containers/device_vector_data.ipp

```diff
@@ -0,0 +1,65 @@
+/** VecMem project, part of the ACTS project (R&D line)
+ *
+ * (c) 2021 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+#pragma once
+
+namespace vecmem {
+
+   template< typename TYPE >
+   VECMEM_HOST_AND_DEVICE
+   device_vector_data< TYPE >::
+   device_vector_data( std::size_t size, pointer ptr )
+   : m_size( size ), m_ptr( ptr ) {
+
+   }
+
+   template< typename TYPE >
+   template< typename ALLOC >
+   VECMEM_HOST
+   device_vector_data< TYPE >::
+   device_vector_data( std::vector< value_type, ALLOC >& vec )
```

Comment:
I don't think this constructor should exist like this, rather it should take its `std::vector` argument as an rvalue reference.

Reply:
I'm not sure what you mean. Unless my C\+\+ vocabulary is completely broken, the constructor **is** receiving an rvalue reference already. The way the two constructors differentiate between which of them is allowed to grab a non-const pointer to the vector's data and which is only allowed to grab a constant one, is by using rvalue references... :confused:

Reply:
Sorry, I should be more clear, what I am suggesting is defining this constructor as `device_vector_data(std::vector<value_type, ALLOC> && vec)` (with two ampersands) to accept rvalue references. If we accept lvalue references like we do now it's possible for the size and content of the original container to go out of sync with the associated `device_vector_data`, which would be bad. Or at least, it could lead to some very unexpected behaviour. If we accept only rvalue references the compiler would protect the user against those mistakes.

Reply:
Thinking about it more there are a few gaps in the specification of `std::vector` that makes this a little difficult... For example, the fact that `.data()` is not guaranteed to return the beginning of the allocation block. Makes taking ownership of it a whole lot harder. Seems like there was a proposal for something like this in [N4359](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2015/n4359.pdf) but it was abandoned. Shame!

Reply:
:confused: We are grabbing these pointers from `vecmem::vector` objects. Which do need to stay alive until the end of the GPU kernel. Since we want to receive the results of the GPU's calculation in them.

Yes, it is very important that host code would stop modifying "host containers" before handing them off to the device code, and only touch these containers again after the device code has finished.

If we would force the host's `vecmem::vector` objects to become invalid once we hand them off to device code, the host code would need to receive a terrible amount of boilerplate. Like:

```c++
vecmem::vector< int > input, output;
vecmem::device_vector_data< const int > device_input( input );
vecmem::device_vector_data< int > device_output( output );
// Hand "device_input" and "device_output" to a GPU kernel, with "input" and "output" having become invalid.
vecmem::vector< int > output_after_gpu_kernel( ? );
```

Since `vecmem::vector` always manages its own memory in the current setup, I have no idea how we would look at the output of a GPU kernel in such a setup in any convenient way. :confused: